backupccl: CI builds panicking with mutex is not write locked

```
panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked

goroutine 627788 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc01406d3b0, 0xa067408, 0xc0079bc040)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x62f5c20, 0x9f7dca0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc01406d3b0, 0xa0674b0, 0xc0096a40f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:231 +0x15d
panic(0x62f5c20, 0x9f7dca0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send.func1(0xc001a51b78, 0xc001a51c30, 0xc004c02000, 0xc001a51c28)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:85 +0x745
panic(0x62f5c20, 0x9f7dca0)
	/usr/local/go/src/runtime/panic.go:971 +0x499
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*Mutex).AssertHeld(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync_race.go:49
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).MaybeGossipSystemConfigRaftMuLocked(0xc004945500, 0xa0674b0, 0xc0096a4150, 0xc003118ea0, 0x5b1688)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_gossip.go:83 +0x865
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleReadWriteLocalEvalResult(0xc004945500, 0xa0674b0, 0xc0096a4150, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_proposal.go:711 +0xc09
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evalAndPropose(0xc004945500, 0xa0674b0, 0xc0096a4150, 0xc007d02dc0, 0xc0137b2ea0, 0x0, 0x0, 0x0, 0x100000001, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:118 +0x1718
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch(0xc004945500, 0xa0674b0, 0xc0096a4150, 0xc007d02dc0, 0xc0137b2ea0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_write.go:149 +0x84e
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc004945500, 0xa0674b0, 0xc0096a4150, 0xc007d02dc0, 0x7155cf8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:376 +0x4a2
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).sendWithRangeID(0xc004945500, 0xa0674b0, 0xc0096a4120, 0x6, 0xc007d02dc0, 0x0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:107 +0x652
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).Send(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_send.go:43
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).Send(0xc004c02000, 0xa0674b0, 0xc0096a40f0, 0x169462bee7428fea, 0x0, 0x100000001, 0x1, 0x0, 0x6, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_send.go:180 +0x5b3
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).Send(0xc007ec6a80, 0xa0674b0, 0xc0096a40f0, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0x6, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/stores.go:191 +0x198
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal.func1(0xa0674b0, 0xc0096a40f0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:901 +0x27e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc01406d3b0, 0xa0674b0, 0xc0096a40f0, 0x6afe160, 0x10, 0xc001a52248, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:324 +0x12a
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc0084a9200, 0xa0674b0, 0xc0096a40f0, 0xc007d02c60, 0x0, 0x0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:889 +0x267
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc0084a9200, 0xa0674b0, 0xc0096a40c0, 0xc007d02c60, 0x5b227c, 0xc00574e510, 0x4a19ac)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:963 +0x118
github.com/cockroachdb/cockroach/pkg/rpc.internalClientAdapter.Batch(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:462
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).sendBatch(0xc00c55fac0, 0xa0674b0, 0xc0096a40c0, 0x1, 0xa0bab28, 0xc01249fc90, 0x0, 0x0, 0x100000001, 0x1, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:195 +0x25d
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*grpcTransport).SendNext(0xc00c55fac0, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x100000001, 0x1, 0x0, 0x6, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport.go:177 +0x25f
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.raceTransport.SendNext(0xa0b3dc8, 0xc00c55fac0, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/transport_race.go:78 +0x443
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendToReplicas(0xc00d08f000, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x6, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1882 +0xdfe
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).sendPartialBatch(0xc00d08f000, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1500 +0x437
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).divideAndSendBatchToRanges(0xc00d08f000, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:1118 +0x1f0a
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*DistSender).Send(0xc00d08f000, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/dist_sender.go:757 +0x77b
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnLockGatekeeper).SendLocked(0xc0036a2960, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_lock_gatekeeper.go:82 +0x1b5
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnMetricRecorder).SendLocked(0xc0036a2928, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_metric_recorder.go:46 +0xfb
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnCommitter).SendLocked(0xc0036a28f8, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_committer.go:200 +0x275
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).sendLockedWithRefreshAttempts(0xc0036a2858, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:240 +0x115
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSpanRefresher).SendLocked(0xc0036a2858, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go:175 +0x36a
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnPipeliner).SendLocked(0xc0036a2788, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_pipeliner.go:285 +0x415
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnSeqNumAllocator).SendLocked(0xc0036a2768, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator.go:105 +0x355
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnHeartbeater).SendLocked(0xc0036a26c8, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_interceptor_heartbeater.go:241 +0x275
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*TxnCoordSender).Send(0xc0036a2500, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord/txn_coord_sender.go:513 +0x575
github.com/cockroachdb/cockroach/pkg/kv.(*DB).sendUsingSender(0xc007ec6900, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:808 +0x175
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Send(0xc007d02580, 0xa0674b0, 0xc00574e510, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:982 +0x228
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).commit(0xc007d02580, 0xa0674b0, 0xc00574e510, 0x27a4f53, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:645 +0x25e
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Commit(0xc007d02580, 0xa0674b0, 0xc00574e510, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:680 +0x10e
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc007d02580, 0xa0674b0, 0xc00574e510, 0xc001a58f30, 0xc007d02580, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:876 +0x4d9
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc007ec6900, 0xa0674b0, 0xc00574e510, 0xc001a58f80, 0x0, 0xc006292170)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:771 +0x125
github.com/cockroachdb/cockroach/pkg/sql/gcjob.deleteDatabaseZoneConfig(0xa0674b0, 0xc00574e510, 0xc007ec6900, 0xc00b1e95d8, 0xc00b1e95d8, 0x0, 0xc000000034, 0xa0674b0, 0xc00574e510)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/descriptor_utils.go:69 +0xb3
github.com/cockroachdb/cockroach/pkg/sql/gcjob.performGC(0xa0674b0, 0xc00574e510, 0xc00c75c700, 0xc006d4b860, 0xc006292140, 0xc00d4e64c0, 0x29)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:74 +0x250
github.com/cockroachdb/cockroach/pkg/sql/gcjob.schemaChangeGCResumer.Resume(0x9695f66c7108001, 0xa0674b0, 0xc00574e510, 0x67b3cc0, 0xc00df7eef8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/gcjob/gc_job.go:167 +0x74f
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine.func1(0xc00eeea1c0, 0xa024470, 0xc00c101410, 0xa0674b0, 0xc00574e510, 0x67b3cc0, 0xc00df7eef8, 0xc001a596e8)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1025 +0x142
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).stepThroughStateMachine(0xc005b80b60, 0xa0674b0, 0xc00574e480, 0x67b3cc0, 0xc00df7eef8, 0xa024470, 0xc00c101410, 0xc0065b6dc0, 0xc005d09c40, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/registry.go:1026 +0xa45
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).runJob(0xc005b80b60, 0xa067408, 0xc0035b3e00, 0xa024470, 0xc00c101410, 0xc0065b6dc0, 0xc005d09c40, 0x7, 0xc00bdd2420, 0x16, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:290 +0x61d
github.com/cockroachdb/cockroach/pkg/jobs.(*Registry).resumeJob.func1(0xa067408, 0xc0079bc040)
	/go/src/github.com/cockroachdb/cockroach/pkg/jobs/adopt.go:222 +0x19b
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2(0xc01406d3b0, 0xa067408, 0xc0079bc040, 0x0, 0x0, 0xc0101deb40)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:442 +0x162
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:434 +0x28e

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000102d80, 0x6b0ffe6, 0x15, 0x71554e8, 0x1)
	/usr/local/go/src/testing/testing.go:1239 +0x610
testing.runTests.func1(0xc000102d80)
	/usr/local/go/src/testing/testing.go:1511 +0xa7
testing.tRunner(0xc000102d80, 0xc000dbfbd8)
	/usr/local/go/src/testing/testing.go:1193 +0x203
testing.runTests(0xc000c1f2f0, 0xc292520, 0x83, 0x83, 0xc036c47b3212c1d5, 0x3463dfe729a, 0xc5d0b40, 0xc0004d3200)
	/usr/local/go/src/testing/testing.go:1509 +0x613
testing.(*M).Run(0xc000b8dc80, 0x0)
	/usr/local/go/src/testing/testing.go:1417 +0x3b4
github.com/cockroachdb/cockroach/pkg/ccl/backupccl.TestMain(0xc000b8dc80)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/backupccl/main_test.go:31 +0x1a9
main.main()
	_testmain.go:317 +0x272
```

Panic on CI build:
https://teamcity.cockroachdb.com/repository/download/Cockroach_UnitTests_Testrace/3216584:id/raw.0.json.txt.tgz!/artifacts/full_output.txt

--- a/pkg/kv/kvserver/replica_application_state_machine.go
+++ b/pkg/kv/kvserver/replica_application_state_machine.go
@@ -1206,7 +1206,7 @@ func (sm *replicaStateMachine) ApplySideEffects(
 	if cmd.IsLocal() {
 		// Handle the LocalResult.
 		if cmd.localResult != nil {
-			sm.r.handleReadWriteLocalEvalResult(ctx, *cmd.localResult)
+			sm.r.handleReadWriteLocalEvalResult(ctx, *cmd.localResult, true /* raftMuHeld */)
 		}
 
 		rejected := cmd.Rejected()

--- a/pkg/kv/kvserver/replica_proposal.go
+++ b/pkg/kv/kvserver/replica_proposal.go
@@ -641,7 +641,9 @@ func addSSTablePreApply(
 	return copied
 }
 
-func (r *Replica) handleReadWriteLocalEvalResult(ctx context.Context, lResult result.LocalResult) {
+func (r *Replica) handleReadWriteLocalEvalResult(
+	ctx context.Context, lResult result.LocalResult, raftMuHeld bool,
+) {
 	// Fields for which no action is taken in this method are zeroed so that
 	// they don't trigger an assertion at the end of the method (which checks
 	// that all fields were handled).
@@ -707,7 +709,19 @@ func (r *Replica) handleReadWriteLocalEvalResult(ctx context.Context, lResult re
 		lResult.MaybeAddToSplitQueue = false
 	}
 
+	// The following three triggers require the raftMu to be held. If a
+	// trigger is present, acquire the mutex if it is not held already.
+	maybeAcquireRaftMu := func() func() {
+		if raftMuHeld {
+			return func() {}
+		}
+		raftMuHeld = true
+		r.raftMu.Lock()
+		return r.raftMu.Unlock
+	}
+
 	if lResult.MaybeGossipSystemConfig {
+		defer maybeAcquireRaftMu()()
 		if err := r.MaybeGossipSystemConfigRaftMuLocked(ctx); err != nil {
 			log.Errorf(ctx, "%v", err)
 		}
@@ -715,6 +729,7 @@ func (r *Replica) handleReadWriteLocalEvalResult(ctx context.Context, lResult re
 	}
 
 	if lResult.MaybeGossipSystemConfigIfHaveFailure {
+		defer maybeAcquireRaftMu()()
 		if err := r.MaybeGossipSystemConfigIfHaveFailureRaftMuLocked(ctx); err != nil {
 			log.Errorf(ctx, "%v", err)
 		}
@@ -722,6 +737,7 @@ func (r *Replica) handleReadWriteLocalEvalResult(ctx context.Context, lResult re
 	}
 
 	if lResult.MaybeGossipNodeLiveness != nil {
+		defer maybeAcquireRaftMu()()
 		if err := r.MaybeGossipNodeLivenessRaftMuLocked(ctx, *lResult.MaybeGossipNodeLiveness); err != nil {
 			log.Errorf(ctx, "%v", err)
 		}

--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -115,7 +115,7 @@ func (r *Replica) evalAndPropose(
 	if proposal.command == nil {
 		intents := proposal.Local.DetachEncounteredIntents()
 		endTxns := proposal.Local.DetachEndTxns(pErr != nil /* alwaysOnly */)
-		r.handleReadWriteLocalEvalResult(ctx, *proposal.Local)
+		r.handleReadWriteLocalEvalResult(ctx, *proposal.Local, false /* raftMuHeld */)
 
 		pr := proposalResult{
 			Reply:              proposal.Local.Reply,

